### DISCUSSION TITLE
use beacon api?

### DISCUSSION THREAD

**Discussion by scottjg** [plausible/analytics#190] [Anything Else]
i love how small the plausible script is, so i'm a little hesistant to suggest more stuff in it, but it would be nice to use the beacon api instead of xhr. that way, you can shoot off events "in the background", as opposed to having them block the page nav. browser support for this is ok but not supported on IE11 which maybe is a dealbreaker for some of your users. in our case, it isn't though.

i guess in theory you could have different scripts served based on the user agent, though i can see why that would be a hassle to maintain.

**Comment by oliverpool** [comment]
> i'm a little hesistant to suggest more stuff in it

I think that the addition can be quite small, and restricted to the `registerOutboundLinkEvents` mode.


> not supported on IE11

One can make a simple `if(window.navigator.sendBeacon)` to detect its support and fallback to the current "delay method".

---

I would be willing to make a PR for the tracking script if the maintainers are ok with it (the `sendBeacon` uses POST, which is already used in the normal case).

**Comment by angelovdev** [comment]
https://github.com/plausible/analytics/blob/master/tracker/src/p.js#L63

```js
if (navigator.sendBeacon) {
    navigator.sendBeacon(plausibleHost + '/api/event', payload)
} else {
...
```

The `sendBeacon` function, [specifically intended to be used for analytics](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon), is supported in all browsers except IE, which the `else` will take care of. If no support for IE11 is required, it can even reduce the code size by eliminating the XMLHttpRequest altogether.

I don't know if this API route is called anywhere else but if this is the only place it's called, it shouldn't be a hard change. 

I am also not sure about the `onreadystatechange` a few lines down.

Maybe @ukutaht can take a look at it.

**Comment by ukutaht** [comment]
Interesting reading on the Beacon API: https://volument.com/blog/sendbeacon-is-broken

Make sure to read the comments for the full context

**Reply by oliverpool** [reply]
> before the unloading of a page

If I understand correctly, the issue is only present when sending the beacon on page unload. Not when sending it upon page load.

Further reading: https://github.com/mdn/content/issues/2692#issuecomment-699526801

**Comment by willfarrell** [comment]
Oops, found this after making a PR (https://github.com/plausible/analytics/pull/1730). Interesting read: https://volument.com/blog/sendbeacon-is-broken (last edited today?). I wonder if this data is still relevant since it was posted 2y ago.

Pulled this link from the nested GitHub thread, as additional reference. https://calendar.perfplanet.com/2020/beaconing-in-practice/

**Comment by igorovic** [comment]
I was wondering if there is a particular reason not to use the beacon api ? `navigator.sendBeacon`
Maybe I am unaware of some caveats.

**Reply by RobertJoonas** [reply]
Internet Explorer doesn't support Beacon. While we do have a `compat` script extension for IE support, bringing Beacon into the mix would complicate the tracking script. We've agreed that the best solution is introducing a newer, more modern script version that would use `navigator.sendBeacon` for every event request. Unfortunately, at this point we cannot make any promises on when we'll be able to work on this. Thanks for understanding!

**Reply by changethe** [reply]
I think most people would much rather get accurate engagement data and just ignore the internet explorer visits, which seems to be a minuscule percentage of visitors anyway. The tradeoff between getting "basic" data for 100% of visitors vs getting detailed data for 99% of visitors seems to be a trivial choice in my eyes.

I hope you get the chance to work on this in the future, consider this a big vote in favor from me!

Good engagement data is the most important part of an analytics tool in my eyes. Basic pageviews/bounce rate stats are just vanity stats I think, and they don't tell me a lot about how engaged users are with my content. The main insight I want to get out of an analytics tool is how users are consuming my content and how much of an impact improvements/changes to the content/site are having on that.

How I understand things, the beacon api seems to be the only solution that can give accurate data for this without any hacky workarounds. 

With the current solutions via custom events, the data seems to be at least somewhat better, but still not 100%. Constant pings or other custom events seem to give very inaccurate data for visit duration, which is the only real engagement metric for content focussed sites.

For users that don't care much about reading time and basically only want to know what sources lead to signups and the like, this metric seems to be not really that important - but for content focussed sites this is the main metric and hence deserves some attention in my eyes.

umami seems to be working on a decent solution for this, maybe there are some insights to be gained from them: https://github.com/umami-software/umami/issues/1470

**Comment by MentalGear** [comment]
What's the status of this? sendBeacon seems to be the most reliable way to go  when used with "visibilitychange" to get a "on leave" package.

**Comment by sjclark** [comment]
Has there been any further progress on sendBeacon? Just trying to figure out if it's planned? Or canned?
